<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


Execute the below code cell to avoid prettytable default error.


In [1]:
!pip install ipython-sql prettytable

import prettytable

prettytable.DEFAULT = 'DEFAULT'

### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [2]:
!pip install pandas
import pandas as pd
import sqlite3

In [3]:
# Load the SQL magic module
conn = sqlite3.connect("Final.db")
cur = conn.cursor()


In [10]:
# Chicago Census Data
census_url="""https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01"""

# Chicago Public Schools
public_schools_url = """https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01"""

# Chicago Crime Data
crime_url = """https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01"""

df_of_census = pd.read_csv(census_url)
df_of_public_schools = pd.read_csv(public_schools_url)
df_of_crime= pd.read_csv(crime_url)


df_of_census.describe()
df_of_public_schools.describe()
df_of_crime.describe()


ID         BEAT    DISTRICT        WARD  \
count  5.330000e+02   533.000000  533.000000  490.000000   
mean   6.228301e+06  1187.037523   11.195122   22.608163   
std    3.027454e+06   660.592263    6.496631   13.080119   
min    2.114900e+04   111.000000    1.000000    1.000000   
25%    3.420516e+06   711.000000    6.000000   12.000000   
50%    6.231712e+06  1111.000000   10.000000   24.000000   
75%    8.875058e+06  1634.000000   16.000000   32.000000   
max    1.127717e+07  2535.000000   25.000000   50.000000   

       COMMUNITY_AREA_NUMBER  X_COORDINATE  Y_COORDINATE        YEAR  \
count             490.000000  5.290000e+02  5.290000e+02   533.00000   
mean               37.542857  1.162876e+06  1.886223e+06  2008.34334   
std                21.409514  1.650197e+04  3.009980e+04     4.94940   
min                 1.000000  1.100658e+06  1.814512e+06  2001.00000   
25%                23.000000  1.151715e+06  1.860454e+06  2004.00000   
50%                30.000000  1.163235e+06  1.891447e+06  2008.00000   
75%                58.000000  1.174618e+06  1.907711e+06  2012.00000   
max                77.000000  1.204126e+06  1.951001e+06  2018.00000   

         LATITUDE   LONGITUDE  
count  529.000000  529.000000  
mean    41.843429  -87.677818  
std      0.082798    0.060073  
min     41.645796  -87.905227  
25%     41.772556  -87.718166  
50%     41.857737  -87.676733  
75%     41.902298  -87.634677  
max     42.021178  -87.528223

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [11]:
df_of_census.to_sql(name='census', con=conn, if_exists='append')
df_of_public_schools.to_sql(name='public_schools', con=conn, if_exists='append')
df_of_crime.to_sql(name='crime', con=conn, if_exists='append')

533

Establish a connection between SQL magic module and the database `FinalDB.db`


In [12]:
!pip install ipython-sql prettytable
import prettytable

# Set the default display format for prettytable to 'DEFAULT' (i.e., a simple table format)
prettytable.DEFAULT = 'DEFAULT'
%load_ext sql
%sql sqlite:///Final.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [24]:
%sql SELECT COUNT(*) FROM crime;

 * sqlite:///Final.db
Done.


[(533,)]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [32]:
df_of_crime.describe()
%sql SELECT LOCATION_DESCRIPTION, COMMUNITY_AREA_NUMBER FROM crime WHERE COMMUNITY_AREA_NUMBER < 11000;

 * sqlite:///Final.db
Done.


[('SMALL RETAIL STORE', 58.0),
 ('OTHER', 23.0),
 ('NURSING HOME/RETIREMENT HOME', 38.0),
 ('RESIDENCE', 46.0),
 ('ALLEY', 66.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 24.0),
 ('STREET', 14.0),
 ('GROCERY FOOD STORE', 38.0),
 ('STREET', 72.0),
 ('SMALL RETAIL STORE', 21.0),
 ('DEPARTMENT STORE', 65.0),
 ('BAR OR TAVERN', 28.0),
 ('RESIDENTIAL YARD (FRONT/BACK)', 49.0),
 ('OTHER', 28.0),
 ('RESIDENCE PORCH/HALLWAY', 26.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 22.0),
 ('OTHER', 48.0),
 ('CTA PLATFORM', 44.0),
 ('VACANT LOT/LAND', 25.0),
 ('SIDEWALK', 25.0),
 ('DEPARTMENT STORE', 8.0),
 ('DEPARTMENT STORE', 32.0),
 ('STREET', 69.0),
 ('GROCERY FOOD STORE', 59.0),
 ('STREET', 68.0),
 ('ATHLETIC CLUB', 14.0),
 ('AIRPORT/AIRCRAFT', 76.0),
 ('STREET', 66.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 73.0),
 ('STREET', 44.0),
 ('OTHER', 32.0),
 ('APARTMENT', 35.0),
 ('SMALL RETAIL STORE', 6.0),
 ('SIDEWALK', 30.0),
 ('STREET', 23.0),
 ('STREET', 68.0),
 ('GROCERY FOOD STORE', 8.0),
 ('RESIDENCE', 43.0),
 ('GROCERY FOOD STORE', 31.0),
 ('VEHICLE NON-COMMERCIAL', 24.0),
 ('STREET', 5.0),
 ('SIDEWALK', 68.0),
 ('SIDEWALK', 22.0),
 ('RESTAURANT', 32.0),
 ('RESIDENCE-GARAGE', 25.0),
 ('NURSING HOME/RETIREMENT HOME', 6.0),
 ('STREET', 68.0),
 ('APARTMENT', 42.0),
 ('STREET', 25.0),
 ('RESIDENCE', 29.0),
 ('RESIDENTIAL YARD (FRONT/BACK)', 23.0),
 ('BAR OR TAVERN', 8.0),
 ('STREET', 25.0),
 ('STREET', 28.0),
 ('VEHICLE NON-COMMERCIAL', 59.0),
 ('DEPARTMENT STORE', 7.0),
 ('STREET', 67.0),
 ('GROCERY FOOD STORE', 6.0),
 ('COMMERCIAL / BUSINESS OFFICE', 32.0),
 ('APARTMENT', 39.0),
 ('OTHER', 77.0),
 ('TAVERN/LIQUOR STORE', 6.0),
 ('STREET', 31.0),
 ('CONSTRUCTION SITE', 61.0),
 ('GAS STATION', 49.0),
 ('STREET', 23.0),
 ('GAS STATION', 25.0),
 ('DRUG STORE', 2.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 63.0),
 ('RESIDENCE', 67.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 69.0),
 ('STREET', 64.0),
 ('APARTMENT', 62.0),
 ('STREET', 66.0),
 ('RESIDENCE', 75.0),
 ('SMALL RETAIL STORE', 65.0),
 ('RESIDENCE PORCH/HALLWAY', 49.0),
 ('DEPARTMENT STORE', 22.0),
 ('STREET', 28.0),
 ('RESIDENCE', 21.0),
 ('STREET', 24.0),
 ('RESIDENCE PORCH/HALLWAY', 1.0),
 ('STREET', 15.0),
 ('CTA BUS', 30.0),
 ('STREET', 71.0),
 ('STREET', 76.0),
 ('STREET', 64.0),
 ('STREET', 16.0),
 ('STREET', 68.0),
 ('STREET', 8.0),
 ('STREET', 77.0),
 ('GROCERY FOOD STORE', 25.0),
 ('STREET', 67.0),
 ('OTHER', 50.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 14.0),
 ('VEHICLE NON-COMMERCIAL', 28.0),
 ('STREET', 73.0),
 ('OTHER', 56.0),
 ('VEHICLE NON-COMMERCIAL', 25.0),
 ('STREET', 51.0),
 ('SIDEWALK', 5.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 25.0),
 ('DRUG STORE', 21.0),
 ('VEHICLE NON-COMMERCIAL', 2.0),
 ('SIDEWALK', 71.0),
 ('APARTMENT', 39.0),
 ('APARTMENT', 30.0),
 ('SCHOOL, PUBLIC, GROUNDS', 43.0),
 ('SIDEWALK', 68.0),
 ('APARTMENT', 61.0),
 ('SCHOOL, PUBLIC, BUILDING', 4.0),
 ('SIDEWALK', 29.0),
 ('SIDEWALK', 43.0),
 ('SIDEWALK', 27.0),
 ('APARTMENT', 24.0),
 ('RESIDENTIAL YARD (FRONT/BACK)', 49.0),
 ('SIDEWALK', 21.0),
 ('SIDEWALK', 23.0),
 ('SIDEWALK', 25.0),
 ('APARTMENT', 8.0),
 ('RESIDENCE', 50.0),
 ('RESIDENCE', 25.0),
 ('CHURCH/SYNAGOGUE/PLACE OF WORSHIP', 73.0),
 ('RESTAURANT', 6.0),
 ('SIDEWALK', 68.0),
 ('APARTMENT', 31.0),
 ('STREET', 42.0),
 ('RESIDENCE PORCH/HALLWAY', 23.0),
 ('RESIDENCE', 25.0),
 ('SCHOOL, PUBLIC, BUILDING', 35.0),
 ('APARTMENT', 29.0),
 ('STREET', 30.0),
 ('RESIDENCE', 19.0),
 ('RESIDENCE', 19.0),
 ('SIDEWALK', 46.0),
 ('RESIDENCE', 27.0),
 ('RESTAURANT', 6.0),
 ('STREET', 7.0),
 ('STREET', 19.0),
 ('RESIDENCE', 29.0),
 ('PARKING LOT/GARAGE(NON.RESID.)', 43.0),
 ('SIDEWALK', 68.0),
 ('DEPARTMENT STORE', 37.0),
 ('SIDEWALK', 68.0),
 ('SIDEWALK', 26.0),
 ('DEPARTMENT STORE', 32.0),
 ('APARTMENT', 6.0),
 ('STREET', 30.0),
 ('SIDEWALK', 67.0),
 ('RESIDENCE', 25.0),
 ('STREET', 69.0),
 ('OTHER COMMERCIAL TRANSPORTATION', 25.0),
 ('SCHOOL, PUBLIC, BUILDING', 25.0),
 ('APARTMENT', 29.0),
 ('RESIDENCE', 16.0),
 ('RESIDENCE PORCH/HALLWAY', 28.0),
 ('CHA APARTMENT', 28.0),


### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [ ]:
%sql SEL

### Problem 4

##### List all kidnapping crimes involving a child?


### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


### Problem 6

##### List the type of schools along with the average safety score for each type.


### Problem 7

##### List 5 community areas with highest % of households below poverty line


### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.



Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
